### PR TITLE
ENH clean up header handling

### LINKED ISSUE #331: change filling of bad characters in header keywords
currently filling with `JUNK____....`

### PR DESCRIPTION
**Pull request by beckermr**
This PR cleans up header handling. It does a few things:

1. When reading hierarch header entries, it allows any ascii chars.
2. It replaces junk it reads with `_` instead of using `JUNK__`.
3. It has tests for reads of total junk.
4. It raises an error for writing header keys with `?`, `*`, or `#` which appear to generate fits files that cannot be read.

superceeds #326
closes #331

### CODE REVIEW

**Comment by beckermr** [comment]
@esheldon @dstndstn I think this has the rest of the work related to header parsing. Appears to pass all tests. If you all can try it on various files you have around to test it, that'd be great!

**Comment by beckermr** [comment]
Also pardon the whitespace changes in C. My editor removes them automatically.

File: fitsio/hdu/base.py
Comment (beckermr) [inline fitsio/hdu/base.py:7]:
```suggestion
```

**Comment by beckermr** [comment]
Note that internall cfitsio appears to choke on using `?` for the weird ascii as opposed to other things like `@` or `<`.

**Comment by beckermr** [comment]
Since these three templare characters write files that cannot be read, I am raising an error if the user attempts to write them.

File: fitsio/test.py
Comment (esheldon) [inline fitsio/test.py:671]: can you add comments here explaining the `_` is expected?

File: fitsio/hdu/base.py
Comment (esheldon) [inline fitsio/hdu/base.py:324]: this is potentially slow in python when writing big headers.  We should test to see if this needs optimization, e.g. compiled regexp or check in C

File: fitsio/test.py
Comment (esheldon) [inline fitsio/test.py:679]: explain why this one went through unchanged

**Review by esheldon — CHANGES_REQUESTED** [review]
This looks good, have some minor comments and requests

**Comment by beckermr** [comment]
@esheldon This is ready for another look!

**Review by esheldon — APPROVED** [review]

**Comment by esheldon** [comment]
@dstndstn do you want to try this out on some of your real world data before we merge?

**Comment by dstndstn** [comment]
Works for my cases!

**Comment by beckermr** [comment]
I have one more idea here. Let's hold off on merging for just a bit.

File: fitsio/fitsio_pywrap.c
Comment (esheldon) [inline fitsio/fitsio_pywrap.c:113]: remove this commented code

**Review by esheldon — CHANGES_REQUESTED** [review]

**Comment by beckermr** [comment]
@dstndstn @esheldon can you all try this out on weird data you have one more time?

The parsing now reads as much as it can without conversion.

File: fitsio/fitsio_pywrap.c
Comment (esheldon) [inline fitsio/fitsio_pywrap.c:4395]: why are these treated differently?  Can you please add a comment?

**Review by esheldon — CHANGES_REQUESTED** [review]
Just to be sure, are we testing a case that would have seg faulted before?

File: fitsio/fitsio_pywrap.c
Comment (beckermr) [reply fitsio/fitsio_pywrap.c:4395]: Yep. I will add a comment. Here is the logic

If a key is HIERARCH, then any ASCII character except `*?#` is allowed. If we don't do any conversion, then we can use the FITS key read functions, otherwise we have to write the value as a string.

If a key is not HIERARCH, then we have more cases. If a key has non-allowed chars for standard fits header keys, then we cannot use the FITS key read functions. So we track that. Also, if a key has non-ascii characters, then we need to convert it before we write it to the python dict. 

All of the if blocks here do that logic. I will add comments.

**Comment by dstndstn** [comment]
This works well for the original Pan-STARRS image I reported problems with (yay!).

**Comment by beckermr** [comment]
ok @esheldon comments added - back to you!

**Comment by beckermr** [comment]
> Just to be sure, are we testing a case that would have seg faulted before?

yes - we are testing stuff that would have segfaulted. I am not sure if it is the same case, but we have weird non-ascii stuff in our junk headers now.

**Review by esheldon — APPROVED** [review]

**Pull request merged**
